File: scripts/generate-test-data.mjs
#!/usr/bin/env node

/**
 * Test Data Generator for Field Worker Scheduler
 * Generates large datasets (100+, 1000+) for performance testing
 */

import fetch from 'node-fetch';

const API_BASE = 'http://localhost:3000/api/trpc';

// Helper to make tRPC calls
async function callTRPC(route, input = {}) {
  const url = new URL(`${API_BASE}/${route}`);
  url.searchParams.append('input', JSON.stringify(input));
  
  const response = await fetch(url.toString());
  const data = await response.json();
  return data.result?.data;
}

// Generate random coordinates around Lagos, Nigeria
function generateCoordinates() {
  const baseLat = 6.5244;
  const baseLng = 3.3792;
  const latOffset = (Math.random() - 0.5) * 0.5; // ~55km radius
  const lngOffset = (Math.random() - 0.5) * 0.5;
  
  return {
    latitude: (baseLat + latOffset).toFixed(6),
    longitude: (baseLng + lngOffset).toFixed(6),
  };
}

// Generate random customer data
function generateCustomer(id, routeId) {
  const coords = generateCoordinates();
  const serviceTypes = ['Delivery', 'Installation', 'Maintenance', 'Inspection', 'Repair'];
  const priorities = ['High', 'Medium', 'Low'];
  
  return {
    name: `Customer ${id}`,
    address: `Address ${id}, Lagos`,
    latitude: coords.latitude,
    longitude: coords.longitude,
    phone: `080${String(Math.floor(Math.random() * 10000000000)).padStart(10, '0')}`,
    email: `customer${id}@example.com`,
    serviceType: serviceTypes[Math.floor(Math.random() * serviceTypes.length)],
    priority: priorities[Math.floor(Math.random() * priorities.length)],
    routeId: routeId,
  };
}

// Generate random route data
function generateRoute(id, workerId, vehicleId, customerCount) {
  const today = new Date();
  const scheduledDate = new Date(today.getTime() + Math.random() * 7 * 24 * 60 * 60 * 1000);
  
  return {
    workerId,
    vehicleId,
    scheduledDate: scheduledDate.toISOString().split('T')[0],
    status: 'pending',
    totalDistance: (Math.random() * 100 + 20).toFixed(2),
    estimatedDuration: Math.floor(Math.random() * 480 + 120), // 2-10 hours
    customerCount,
  };
}

// Main generation function
async function generateTestData(customerCount) {
  console.log(`\n🚀 Generating test data for ${customerCount} customers...`);
  
  try {
    // Get or create workers
    console.log('📋 Creating workers...');
    const workers = [
      { name: 'John Doe', phone: '08012345678', pin: '1234' },
      { name: 'Jane Smith', phone: '08087654321', pin: '5678' },
      { name: 'Mike Johnson', phone: '08098765432', pin: '9012' },
    ];
    
    const workerIds = [];
    for (const worker of workers) {
      try {
        const result = await callTRPC('fieldWorker.createWorker', worker);
        if (result?.id) {
          workerIds.push(result.id);
          console.log(`  ✓ Created worker: ${worker.name} (ID: ${result.id})`);
        }
      } catch (e) {
        console.log(`  ⚠ Worker ${worker.name} might already exist`);
      }
    }
    
    if (workerIds.length === 0) {
      console.log('  ℹ Using existing workers...');
      // Fallback to first 3 workers
      workerIds.push(1, 2, 3);
    }
    
    // Create vehicles
    console.log('🚗 Creating vehicles...');
    const vehicles = [
      { plateNumber: 'LG-001-ABC', capacity: 500, status: 'available' },
      { plateNumber: 'LG-002-DEF', capacity: 500, status: 'available' },
      { plateNumber: 'LG-003-GHI', capacity: 500, status: 'available' },
    ];
    
    const vehicleIds = [];
    for (const vehicle of vehicles) {
      try {
        const result = await callTRPC('fieldWorker.createVehicle', vehicle);
        if (result?.id) {
          vehicleIds.push(result.id);
          console.log(`  ✓ Created vehicle: ${vehicle.plateNumber} (ID: ${result.id})`);
        }
      } catch (e) {
        console.log(`  ⚠ Vehicle ${vehicle.plateNumber} might already exist`);
      }
    }
    
    if (vehicleIds.length === 0) {
      vehicleIds.push(1, 2, 3);
    }
    
    // Create routes and customers
    console.log(`📍 Creating ${customerCount} customers in routes...`);
    const customersPerRoute = Math.ceil(customerCount / workerIds.length);
    let totalCustomers = 0;
    let totalRoutes = 0;
    
    for (let w = 0; w < workerIds.length; w++) {
      const workerId = workerIds[w];
      const vehicleId = vehicleIds[w % vehicleIds.length];
      const routeCustomerCount = Math.min(customersPerRoute, customerCount - totalCustomers);
      
      if (routeCustomerCount <= 0) break;
      
      // Create route
      const routeData = generateRoute(totalRoutes + 1, workerId, vehicleId, routeCustomerCount);
      try {
        const routeResult = await callTRPC('fieldWorker.createRoute', routeData);
        if (routeResult?.id) {
          const routeId = routeResult.id;
          totalRoutes++;
          console.log(`  ✓ Created route ${totalRoutes} for worker ${workerId} with ${routeCustomerCount} customers`);
          
          // Create customers for this route
          for (let c = 0; c < routeCustomerCount; c++) {
            const customerData = generateCustomer(totalCustomers + 1, routeId);
            try {
              const customerResult = await callTRPC('fieldWorker.createCustomer', customerData);
              if (customerResult?.id) {
                totalCustomers++;
                if (totalCustomers % 100 === 0) {
                  console.log(`    ✓ Created ${totalCustomers} customers...`);
                }
              }
            } catch (e) {
              console.error(`    ✗ Failed to create customer: ${e.message}`);
            }
          }
        }
      } catch (e) {
        console.error(`  ✗ Failed to create route: ${e.message}`);
      }
    }
    
    console.log(`\n✅ Test data generation complete!`);
    console.log(`   - Workers: ${workerIds.length}`);
    console.log(`   - Vehicles: ${vehicleIds.length}`);
    console.log(`   - Routes: ${totalRoutes}`);
    console.log(`   - Customers: ${totalCustomers}`);
    console.log(`\n📊 Performance Testing:`);
    console.log(`   - Test with ${totalCustomers} customers`);
    console.log(`   - Run route optimization to measure performance`);
    
  } catch (error) {
    console.error('❌ Error generating test data:', error);
    process.exit(1);
  }
}

// Parse command line arguments
const args = process.argv.slice(2);
const customerCount = parseInt(args[0]) || 100;

if (customerCount < 10) {
  console.error('❌ Customer count must be at least 10');
  process.exit(1);
}

generateTestData(customerCount);

File: scripts/generate_test_data.py
#!/usr/bin/env python3

import os
import sys
import random
import mysql.connector
from datetime import datetime, timedelta

customer_count = int(sys.argv[1]) if len(sys.argv) > 1 else 100

db_url = os.environ.get('DATABASE_URL', '')
if not db_url:
    print("❌ DATABASE_URL not set")
    sys.exit(1)

try:
    from urllib.parse import urlparse
    parsed = urlparse(db_url)
    config = {
        'host': parsed.hostname,
        'user': parsed.username,
        'password': parsed.password,
        'database': parsed.path.lstrip('/'),
        'port': parsed.port or 3306,
    }
except:
    print("❌ Invalid DATABASE_URL format")
    sys.exit(1)

def generate_coordinates():
    base_lat = 6.5244
    base_lng = 3.3792
    lat_offset = (random.random() - 0.5) * 0.5
    lng_offset = (random.random() - 0.5) * 0.5
    return round(base_lat + lat_offset, 6), round(base_lng + lng_offset, 6)

def seed_test_data(customer_count):
    print(f"\n🚀 Seeding test data for {customer_count} customers...")
    
    try:
        conn = mysql.connector.connect(**config)
        cursor = conn.cursor()
        
        print("📋 Creating workers...")
        workers_data = [
            ('Test Worker 1', 'worker1@test.com', '08012345678'),
            ('Test Worker 2', 'worker2@test.com', '08087654321'),
            ('Test Worker 3', 'worker3@test.com', '08098765432'),
        ]
        
        worker_ids = []
        for name, email, phone in workers_data:
            try:
                cursor.execute(
                    "INSERT INTO workers (name, email, phone, status, createdAt, updatedAt) "
                    "VALUES (%s, %s, %s, 'active', NOW(), NOW())",
                    (name, email, phone)
                )
                conn.commit()
                worker_id = cursor.lastrowid
                worker_ids.append(worker_id)
                print(f"  ✓ Worker {name} (ID: {worker_id})")
            except Exception as e:
                print(f"  ✗ Error creating worker {name}: {e}")
        
        print("🚗 Creating vehicles...")
        vehicles_data = [
            ('Test Vehicle 1', 'TEST-001', 500),
            ('Test Vehicle 2', 'TEST-002', 500),
            ('Test Vehicle 3', 'TEST-003', 500),
        ]
        
        vehicle_ids = []
        for name, plate, capacity in vehicles_data:
            try:
                cursor.execute(
                    "INSERT INTO vehicles (name, plateNumber, capacity, status, startLatitude, startLongitude, createdAt, updatedAt) "
                    "VALUES (%s, %s, %s, 'available', '6.5244', '3.3792', NOW(), NOW())",
                    (name, plate, capacity)
                )
                conn.commit()
                vehicle_id = cursor.lastrowid
                vehicle_ids.append(vehicle_id)
                print(f"  ✓ Vehicle {plate} (ID: {vehicle_id})")
            except Exception as e:
                print(f"  ✗ Error creating vehicle {plate}: {e}")
        
        if not vehicle_ids or not worker_ids:
            print("❌ Failed to create workers or vehicles")
            return
        
        print(f"📍 Creating {customer_count} customers in routes...")
        
        customers_per_route = (customer_count + len(worker_ids) - 1) // len(worker_ids)
        total_customers = 0
        total_routes = 0
        
        for w_idx, worker_id in enumerate(worker_ids):
            vehicle_id = vehicle_ids[w_idx % len(vehicle_ids)]
            route_customer_count = min(customers_per_route, customer_count - total_customers)
            
            if route_customer_count <= 0:
                break
            
            scheduled_date = (datetime.now() + timedelta(days=w_idx)).date()
            total_distance = str(round(random.random() * 100 + 20, 2))
            estimated_duration = str(random.randint(120, 600))
            
            try:
                cursor.execute(
                    "INSERT INTO routes (workerId, vehicleId, scheduledDate, status, totalDistance, estimatedDuration, createdAt, updatedAt) "
                    "VALUES (%s, %s, %s, 'pending', %s, %s, NOW(), NOW())",
                    (worker_id, vehicle_id, scheduled_date, total_distance, estimated_duration)
                )
                conn.commit()
                route_id = cursor.lastrowid
                total_routes += 1
                print(f"  ✓ Created route {total_routes} for worker {worker_id} with {route_customer_count} customers")
                
                customer_values = []
                for c in range(route_customer_count):
                    lat, lng = generate_coordinates()
                    service_types = ['Delivery', 'Installation', 'Maintenance', 'Inspection', 'Repair']
                    priorities = ['high', 'medium', 'low']
                    
                    customer_values.append((
                        f"Test Customer {total_customers + c + 1}",
                        f"Address {total_customers + c + 1}, Lagos",
                        str(lat),
                        str(lng),
                        f"080{random.randint(1000000000, 9999999999)}",
                        f"customer{total_customers + c + 1}@test.com",
                        random.choice(service_types),
                        random.choice(priorities),
                    ))
                
                if customer_values:
                    cursor.executemany(
                        "INSERT INTO customers (name, address, latitude, longitude, phone, email, serviceType, priority, createdAt, updatedAt) "
                        "VALUES (%s, %s, %s, %s, %s, %s, %s, %s, NOW(), NOW())",
                        customer_values
                    )
                    conn.commit()
                    
                    cursor.execute("SELECT id FROM customers ORDER BY id DESC LIMIT %s", (route_customer_count,))
                    customer_ids = [row[0] for row in cursor.fetchall()]
                    customer_ids.reverse()
                    
                    route_customer_values = [(route_id, cid) for cid in customer_ids]
                    cursor.executemany(
                        "INSERT INTO routeCustomers (routeId, customerId) VALUES (%s, %s)",
                        route_customer_values
                    )
                    conn.commit()
                    
                    total_customers += len(customer_values)
                    
                    if total_customers % 50 == 0:
                        print(f"    ✓ Created {total_customers} customers...")
            
            except Exception as e:
                print(f"  ✗ Error creating route: {e}")
        
        print(f"\n✅ Test data seeding complete!")
        print(f"   - Workers: {len(worker_ids)}")
        print(f"   - Vehicles: {len(vehicle_ids)}")
        print(f"   - Routes: {total_routes}")
        print(f"   - Customers: {total_customers}")
        
        cursor.close()
        conn.close()
        
    except Exception as error:
        print(f"❌ Error: {error}")
        sys.exit(1)

if __name__ == '__main__':
    seed_test_data(customer_count)
